write-fonts/src/pens.rs

```diff
@@ -0,0 +1,369 @@
+}
+
+impl<'a, T: Pen> TransformPen<'a, T> {
+    pub fn new(inner_pen: &'a mut T, transform: Affine) -> TransformPen<'a, T> {
+        TransformPen {
+            inner_pen,
+            transform,
+        }
+    }
+
+    fn map_point(&self, x: f32, y: f32) -> (f32, f32) {
+        let pt = self.transform
+            * Point {
+                x: x as f64,
+                y: y as f64,
+            };
+        (pt.x as f32, pt.y as f32)
+    }
+}
+
+impl<'a, T: Pen> Pen for TransformPen<'a, T> {
+    fn move_to(&mut self, x: f32, y: f32) {
+        let (x, y) = self.map_point(x, y);
+        self.inner_pen.move_to(x, y);
+    }
+
+    fn line_to(&mut self, x: f32, y: f32) {
+        let (x, y) = self.map_point(x, y);
+        self.inner_pen.line_to(x, y);
+    }
+
+    fn quad_to(&mut self, cx0: f32, cy0: f32, x: f32, y: f32) {
+        let (cx0, cy0) = self.map_point(cx0, cy0);
+        let (x, y) = self.map_point(x, y);
+        self.inner_pen.quad_to(cx0, cy0, x, y);
+    }
+
+    fn curve_to(&mut self, cx0: f32, cy0: f32, cx1: f32, cy1: f32, x: f32, y: f32) {
+        let (cx0, cy0) = self.map_point(cx0, cy0);
+        let (cx1, cy1) = self.map_point(cx1, cy1);
+        let (x, y) = self.map_point(x, y);
+        self.inner_pen.curve_to(cx0, cy0, cx1, cy1, x, y);
+    }
+
+    fn close(&mut self) {
+        self.inner_pen.close();
+    }
+}
+
+pub struct BezPathPen {
+    path: BezPath,
+}
+
+fn pt(x: f32, y: f32) -> Point {
+    Point {
+        x: x as f64,
+        y: y as f64,
+    }
+}
+
+impl BezPathPen {
+    pub fn new() -> BezPathPen {
+        BezPathPen {
+            path: BezPath::new(),
+        }
+    }
+
+    pub fn into_inner(self) -> BezPath {
+        self.path
+    }
+}
+
+impl Default for BezPathPen {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Pen for BezPathPen {
+    fn move_to(&mut self, x: f32, y: f32) {
+        self.path.move_to(pt(x, y))
+    }
+
+    fn line_to(&mut self, x: f32, y: f32) {
+        self.path.line_to(pt(x, y))
+    }
+
+    fn quad_to(&mut self, cx0: f32, cy0: f32, x: f32, y: f32) {
+        self.path.quad_to(pt(cx0, cy0), pt(x, y));
+    }
+
+    fn curve_to(&mut self, cx0: f32, cy0: f32, cx1: f32, cy1: f32, x: f32, y: f32) {
+        self.path.curve_to(pt(cx0, cy0), pt(cx1, cy1), pt(x, y));
+    }
+
+    fn close(&mut self) {
+        self.path.close_path();
+    }
+}
+
+#[derive(Debug)]
+pub enum ContourReversalError {
+    InvalidFirstCommand(PenCommand),
+}
+
+/// Buffers commands until a close is seen, then plays in reverse on inner pen.
+///
+/// Reverses the winding direction of the contour. Keeps the first point unchanged.
+///
+/// <https://github.com/fonttools/fonttools/blob/78e10d8b42095b709cd4125e592d914d3ed1558e/Lib/fontTools/pens/reverseContourPen.py#L8>
+pub struct ReverseContourPen<'a, T: Pen> {
+    inner_pen: &'a mut T,
+    pending: Vec<PenCommand>,
+}
+
+fn end(subpath_start_x: f32, subpath_start_y: f32, subpath_cmd: PenCommand) -> (f32, f32) {
```

Comment:
This seems like it might be useful as a public `end_point` method on `PenCommand` that returns an `Option<(f32, f32)>` yielding `None` in the case of the close command.

Reply:
👍  I thought exactly the same

Reply:
Good call, will update so.